src/lib/comms/sol-lwm2m-server.c

```diff
@@ -812,67 +813,104 @@ observer_entry_del_monitor(struct observer_entry *entry,
 }
 
 SOL_API struct sol_lwm2m_server *
-sol_lwm2m_server_new(uint16_t coap_port, uint16_t dtls_port,
-    enum sol_lwm2m_security_mode sec_mode, struct sol_vector *known_psks)
+sol_lwm2m_server_new(uint16_t coap_port, uint16_t num_sec_modes, ...)
 {
     struct sol_lwm2m_server *server;
     struct sol_network_link_addr servaddr_coap = { .family = SOL_NETWORK_FAMILY_INET6,
                                                    .port = coap_port };
     struct sol_network_link_addr servaddr_dtls = { .family = SOL_NETWORK_FAMILY_INET6,
-                                                   .port = dtls_port };
-    int r;
-    const char *sec_mode_str;
-    struct sol_lwm2m_security_psk *psk, *given_psk;
+                                                   .port = SOL_LWM2M_DEFAULT_SERVER_PORT_DTLS };
+    int r, dtls_port;
+    struct sol_lwm2m_security_psk **known_psks = NULL;
+    struct sol_lwm2m_security_rpk *my_rpk = NULL;
+    struct sol_blob **known_pub_keys = NULL;
+    enum sol_lwm2m_security_mode *sec_modes = NULL;
+    enum sol_socket_dtls_cipher *cipher_suites = NULL;
     uint16_t i;
+    va_list ap;
 
     SOL_LOG_INTERNAL_INIT_ONCE;
 
-    switch (sec_mode) {
-    case SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY:
-        SOL_NULL_CHECK(known_psks, NULL);
-        sec_mode_str = "Pre-Shared Key";
-        break;
-    case SOL_LWM2M_SECURITY_MODE_RAW_PUBLIC_KEY:
-        sec_mode_str = "Raw Public Key";
-        SOL_WRN("Raw Public Key security mode is not supported yet.");
-        return NULL;
-    case SOL_LWM2M_SECURITY_MODE_CERTIFICATE:
-        sec_mode_str = "Certificate";
-        SOL_WRN("Certificate security mode is not supported yet.");
-        return NULL;
-    case SOL_LWM2M_SECURITY_MODE_NO_SEC:
-        sec_mode_str = "NoSec";
-        SOL_DBG("Using NoSec Security Mode (No DTLS).");
-        break;
-    default:
-        SOL_WRN("Unknown DTLS Security Mode: %d", sec_mode);
-        return NULL;
+    va_start(ap, num_sec_modes);
+
+    if (num_sec_modes > 0) {
```

Comment:
Same comments as found in the bootstrap.
* Do it in one pass.
* Do not allow the user to pass repeated security modes.
* remember to ref all the blobs (psks and rpks)
* Do not server->known_psks = known_psks.